src/controllers/listings.controller.js

```diff
@@ -6,106 +6,119 @@ import {
   toggleListingStatus,
   getInactiveListings
 } from '../servises/listings.service.js';
-import { validateListingData } from '../validators/listing.validator.js';
-import { AppError } from '../utils/errors.js';
-
-export const createListingAction = async (req, res, next) => {
-  try {
-    const { title, description, price, category } = req.body;
-    const userId = req.user?.id;
-
-    const validationError = validateListingData({ title, description, price, category, userId });
-    if (validationError) {
-      throw new AppError(validationError, 400);
-    }
-
-    const newListing = await createListing({ title, description, price, category, userId });
-    res.status(201).json(newListing);
-  } catch (error) {
-    next(error);
+import { validateRequiredFields, validateUserId, validateListingData } from '../utils/validators.js';
+import { asyncHandler } from '../middleware/asyncHandler.js';
+import { ValidationError, NotFoundError, ForbiddenError } from '../utils/errors.js';
+
+export const createListingAction = asyncHandler(async (req, res) => {
+  const { title, description, price, category } = req.body;
+  const userId = req.user?.id;
+
+  if (!userId) {
```

Comment:
All these validations could be places into one function. So it would be easier to use here. Just call one function `validateListingData` and it will check userId, all required fields, and that listing is correct. Just one line of code to call the validation function, and all of the logic is inside of this function.